src/sage/topology/simplicial_complex.py

```diff
@@ -4802,7 +4810,110 @@ def intersection(self, other):
             F = F + [s for s in self.faces()[k] if s in other.faces()[k]]
         return SimplicialComplex(F)
 
+    def bigraded_betti_numbers(self, base_ring=ZZ):
+        r"""
+        Return a dictionary of the bigraded Betti numbers of ``self``, 
+        with keys `(-i, 2j)`.
+
+        .. SEEALSO::
+
+            See :meth:`bigraded_betti_number` for more information.
+
+        EXAMPLES::
+
+            sage: X = SimplicialComplex([[0,1],[1,2],[1,3],[2,3]])
+            sage: Y = SimplicialComplex([[1,2,3],[1,2,4],[3,5],[4,5]])
+            sage: X.bigraded_betti_numbers()
+            {(0, 0): 1, (-1, 4): 2, (-1, 6): 1, (-2, 6): 1, (-2, 8): 1}
+            sage: Y.bigraded_betti_numbers(base_ring=QQ)
+            {(0, 0): 1, (-1, 4): 3, (-2, 6): 1, (-2, 8): 2, (-3, 10): 1}
+        """
+        if self._bbn is not None and 0 in self._bbn and base_ring in self._bbn[0]:
```

Comment:
```suggestion
        if self._bbn is not None and base_ring in self._bbn:
```

Reply:
I fail to see why this will work. For example, if we were to call `X.bigraded_betti_number(-3, 12)`, then `X._bbn` will have a `key:value` pair `Integer Ring: {(-3, 12): 25}` (I added this to `bigraded_betti_number()` in order to cache calls for single values as well). In that case, if we only check `base_ring in self._bbn`, the method `bigraded_betti_numbers()` will return `{(-3, 12): 25}` (if `X.bigraded_betti_number(-3, 12)` was called prior to `X.bigraded_betti_numbers()`), and not the entire dictionary of all bigraded Betti numbers. The idea behind using `0` as a key is to associate with it the set of all base rings for which `bigraded_betti_numbers()` was called, if a base ring is not in that set, then we have to compute the entire dictionary (the key value `0` is probably poorly chosen). Perhaps I am missing something.

Reply:
Oh, I see what you were doing. I would put that as a separate `dict`, but you could just put some comments in the code explaining this. It is really opaque otherwise.

Reply:
Adding another `dict` did occur to me, but I did not want to add any unnecessary attributes, although it is a bit "messy" like this. I will comment the code as best as I can to explain the structure of `_bbn`.

Reply:
An idea just occurred to me. How about we set `self._bbn = {}` in `__init__`. Then we wouldn't have to check in the methods whether `self._bbn is not None`. This would make the code a lot more readable, I think.

Reply:
> Adding another `dict` did occur to me, but I did not want to add any unnecessary attributes, although it is a bit "messy" like this. I will comment the code as best as I can to explain the structure of `_bbn`.

I would have to think somewhat hard about what is going on to understand it, even with comments. Another (hidden) attribute is cheap and more natural code.

Reply:
> An idea just occurred to me. How about we set `self._bbn = {}` in `__init__`? Then we wouldn't have to check in the methods whether `self._bbn is not None`. This would make the code a lot more readable, I think.

Considering how you are now doing it, that would be reasonable.

Reply:
> I would have to think somewhat hard about what is going on to understand it, even with comments. Another (hidden) attribute is cheap and more natural code.

I think that I will add another attribute, it does make more sense. In that case, do you think that we should initialize them to  `{}` instead of `None`, as it would still make the code more compact?

Reply:
Yes, I would also initialize it to an empty dict `{}`.